OpenRA.Mods.Common/Traits/Player/ParallelProductionQueue.cs

```diff
@@ -0,0 +1,82 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class ParallelProductionQueueInfo : ProductionQueueInfo
+	{
+		public override object Create(ActorInitializer init) { return new ParallelProductionQueue(init, this); }
+	}
+
+	public class ParallelProductionQueue : ProductionQueue
+	{
+		[Sync] public new bool CurrentDone { get { return Queue.Find(i => i.Done) != null; } }
+
+		public ParallelProductionQueue(ActorInitializer init, ProductionQueueInfo info) : base(init, init.Self.Owner.PlayerActor, info)
+		{
+		}
+
+		protected override void ProgressQueue()
+		{
+			for (var i = 0; i < Queue.Count;)
+			{
+				var first = Queue[0];
+
+				var before = first.RemainingTime;
+				first.Tick(playerResources);
+
+				foreach (var item in Queue.FindAll(a => a.Item == first.Item))
+				{
+					++i;
+					Queue.Remove(item);
+					Queue.Add(item);
```

Comment:
Reordering while looping through the collection is still bad, you could be ticking the same item twice for example

Reply:
If you are sure this is how you want to do it then you should make it into a foreach loop and do Queue.ToArray() so you are not looping through the collection you are modifying.

Reply:
this code isn't?
```
for (var i = 0; i < Queue.Count;)
{
    var first = Queue[0];
```

Reply:
ohh wait nvm, I didn't see that it was using Queue[0], my brain thought it was Queue[i] my misake >_>